packages/core/notifications/core-notifications-browser-internal/src/toasts/deduplicate_toasts.tsx

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { Toast } from '@kbn/core-notifications-browser';
+import { EuiNotificationBadge } from '@elastic/eui';
+import React from 'react';
+
+const TOAST_TIMEOUT_LONG = 6e6;
+
+/**
+ * Collects toast messages to groups based on the `getKeyOf` function,
+ * then represents every group of message with a single toast
+ * @param allToasts
+ * @param dismissToast
+ */
+export function deduplicateToasts(allToasts: Toast[], dismissToast: (id: string) => void) {
+  const toastGroups = groupByKey(allToasts);
+
+  const distinctToasts: Toast[] = [];
+  for (const toastGroup of Object.values(toastGroups)) {
+    if (toastGroup.length === 1) {
+      distinctToasts.push(toastGroup[0]);
+    } else {
+      distinctToasts.push(
+        mergeToasts(toastGroup, () => toastGroup.forEach((t) => dismissToast(t.id)))
+      );
+    }
+  }
+
+  return distinctToasts;
+}
+
+/**
+ * Extracts a key from a toast message
+ * these keys decide what makes between different toasts, and which ones should be merged
+ * Ideally we want different (header+text) combinations to appear as separate toast messages
+ * @param toast
+ */
+function getKeyOf(toast: Toast): string {
+  if (isString(toast.title) && isString(toast.text)) {
+    return toast.title + ' ' + toast.text;
+  } else if (isString(toast.text)) {
+    return toast.text;
+  } else if (isString(toast.title)) {
+    return toast.title;
+  } else {
+    // Both title & text is missing or a mount function
+    return 'KEY_' + toast.id.toString();
+  }
+}
+
+function isString(a: string | any): a is string {
+  return typeof a === 'string';
+}
+
+function groupByKey(allToasts: Toast[]) {
+  const toastGroups: Record<string, Toast[]> = {};
+  for (const toast of allToasts) {
+    const key = getKeyOf(toast);
+
+    if (!toastGroups[key]) {
+      toastGroups[key] = [toast];
+    } else {
+      toastGroups[key].push(toast);
+    }
+  }
+  return toastGroups;
+}
+
+function mergeToasts(toasts: Toast[], closeAll: () => void) {
+  const firstElement = toasts[0];
+  const key = getKeyOf(firstElement);
+  return {
+    ...firstElement,
+    id: key,
+    text: firstElement.text,
+    toastLifeTimeMs: TOAST_TIMEOUT_LONG,
```

Comment:
I think this means that grouped toasts will never auto dismiss. We probably should think of something. Maybe just use the timeout of the first toast? Or pick the longest timeout? I know it won't be ideal, but it probably would be better than making them indefinite.

Reply:
The only edge case I think of is, for example: 

1st toast comes with timeout 1s,
2nd toast comes after 500ms 

we merge them and technically this is still first toast displayed and it will be closed in 500ms. So technically 2nd toast has been shown for 500ms instead of 1s. Not sure if we should worry about this or if this is something we can easily fix.

Reply:
What I think is that if there are multiple notifications with the same message+title are in the system, they're probably emitted from the same line in the code, probably having the same timeout setting, so I don't see this being realistic.

Reply:
I mean that two same toasts could emit with a delay between them. Then the second toast will hide earlier than needed

Reply:
I see - well, in that case I don't know what would be the expected behaviour :) For now, let's try it like you suggested, or later fall back to non-disappearing merged toasts.